codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,467 @@
+        return http2Handler;
+    }
+
+    /**
+     * Save context and load any dependencies.
+     */
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = ctx;
+        ctx.pipeline().addBefore(ctx.executor(), ctx.name(), null, http2Handler);
+        this.http2HandlerCtx = ctx.pipeline().context(http2Handler);
+    }
+
+    /**
+     * Clean up any dependencies.
+     */
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+      ctx.pipeline().remove(http2Handler);
+    }
+
+    /**
+     * Handles the cleartext HTTP upgrade event. If an upgrade occurred, sends a simple response via
+     * HTTP/2 on stream 1 (the stream specifically reserved for cleartext HTTP upgrade).
+     */
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (!(evt instanceof UpgradeEvent)) {
+            super.userEventTriggered(ctx, evt);
+            return;
+        }
+
+        UpgradeEvent upgrade = (UpgradeEvent) evt;
+        ctx.fireUserEventTriggered(upgrade.retain());
+        try {
+            Http2Stream stream = http2Handler.connection().stream(Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
+            // TODO: improve handler/stream lifecycle so that stream isn't active before handler added.
+            // The stream was already made active, but ctx may have been null so it wasn't initialized.
+            new ConnectionListener().onStreamActive(stream);
+            upgrade.upgradeRequest().headers().setInt(
+                HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
+            new InboundHttpToHttp2Adapter(http2Handler.connection(), http2Handler.decoder().frameListener())
+                .channelRead(ctx, upgrade.upgradeRequest().retain());
+        } finally {
+            upgrade.release();
+        }
+    }
+
+    /**
+     * Processes all {@link Http2Frame}s. {@link Http2StreamFrame}s may only originate in child
+     * streams.
+     */
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+            throws Exception {
+        if (!(msg instanceof Http2Frame)) {
+            super.write(ctx, msg, promise);
+            return;
+        }
+        try {
+            boolean unexpected = false;
+            if (msg instanceof Http2StreamFrame) {
+                Object streamObject = ((Http2StreamFrame) msg).stream();
+                if (!(streamObject instanceof Http2StreamChannel)) {
+                    // TODO: swap to failing the promise. Today, however, the promise is likely
+                    // being ignored.
+                    throw new IllegalArgumentException(
+                        "Http2StreamFrame.stream() must be an Http2StreamChannel: " + streamObject);
+                }
+                int streamId = ((Http2StreamChannel) streamObject).stream.id();
```

Comment:
This may not line up with long term plans, and maybe exposing types that you ultimately don't want exposed. But instead of a `instanceof` if block that must be updated when new frame types are created/supported what about adding a method to `Http2StreamFrame`:

```java
public interface Http2StreamFrame {
  void writeFrame(ChannelHandlerContext ctx, Http2FrameWriter writer);
}
```